### PR TITLE
S6388 should require all conditions to be met to raise on osDisk or dataDisks 

### PR DESCRIPTION
**Pull request by GabinL21**

### CODE REVIEW

**Review by jonas-wielage-sonarsource — CHANGES_REQUESTED** [review]
Nice and clean implementation, I like it!
I have some feedback though, that needs to be addressed before merging.

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/UnencryptedCloudServicesCheck.java [hunk 1]
```diff
@@ -44,18 +44,9 @@ public class UnencryptedCloudServicesCheck extends AbstractArmResourceCheck {
...
-        .forEach(UnencryptedCloudServicesCheck::checkForDiskEncryptionSet));
+    register("Microsoft.Compute/virtualMachines", UnencryptedCloudServicesCheck::checkVirtualMachineDataDisks);
 
-    register("Microsoft.Compute/virtualMachines",
-      resource -> resource.object("storageProfile").object("osDisk").property("encryptionSettings")
-        .reportIf(isFalse(), UNENCRYPTED_MESSAGE)
-        .reportIfAbsent(FORMAT_OMITTING));
+    register("Microsoft.Compute/virtualMachines", UnencryptedCloudServicesCheck::checkVirtualMachineOsDisk);
 
     register("Microsoft.Compute/virtualMachineScaleSets",
```
[the remark this answers is not in the document]
Comment (GabinL21) [reply iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/UnencryptedCloudServicesCheck.java:51]: Indeed! It should be covered by SONARIAC-1100, so I made no changes for now (and to prevent having an even bigger PR).

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/tree/impl/bicep/StringLiteralImpl.java [hunk 2]
```diff
@@ -43,4 +43,9 @@ public String value() {
   public List<Tree> children() {
     return List.of(token);
   }
+
+  @Override
+  public String toString() {
+    return token.toString();
+  }
```
[the remark this answers is not in the document]
Comment (GabinL21) [reply iac-extensions/arm/src/main/java/org/sonar/iac/arm/tree/impl/bicep/StringLiteralImpl.java:50]: Yes, almost every AST element doesn't have one, and it's hard to debug without it (in my sense, at least).
When running unit tests in debug with a breakpoint, you have to expand every dropdown to see which test is currently running, which element corresponds to which part of the test, etc.

It now looks like this:
image


There is no downside to adding `toString`methods. The only one I thought about was with big unit test files. It could generate long strings in debug. But in that case, it fails pretty quickly with an `OutOfMemory`exception; I tested it. I'm trying to add those methods as I need them in my PR since I'm adding them locally anyway :)

File: iac-extensions/arm/src/test/java/org/sonar/iac/arm/tree/impl/bicep/ObjectExpressionImplTest.java [hunk 3]
```diff
@@ -54,7 +53,11 @@ void shouldParseMinimumObjectExpression() {
 
   @Test
   void shouldParseObjectExpressionWithOneProperty() {
-    String code = code("{\n key1: value1\n 'key2': value2\n}\n");
+    String code = """
+      {
+        key1: value1
+        'key2': value2
+      }""";
```
[the remark this answers is not in the document]
Comment (GabinL21) [reply iac-extensions/arm/src/test/java/org/sonar/iac/arm/tree/impl/bicep/ObjectExpressionImplTest.java:60]: Was wondering about this format! Shouldn't it be covered by Spotless? Not sure about how much control we have over it

[the remark this answers is not in the document]
Comment (jonas-wielage-sonarsource) [reply iac-extensions/arm/src/test/java/org/sonar/iac/arm/tree/impl/bicep/ObjectExpressionImplTest.java:60]: No, those are two separate strings:
`}""";` corresponds to `"}"`, whereas my suggestion corresponds to `"}\n"`.
In the end it doesn't matter a lot, I just noticed that it changed from the previous implementation and wasn't sure you're aware of the difference.

File: its/ruling/src/integrationTest/resources/expected/azureresourcemanager/azureresourcemanager-S6388.json [hunk 4]
```diff
@@ -6,22 +6,28 @@
 "azureresourcemanager-project:ruling/src/integrationTest/resources/sources/azureresourcemanager/rules/sas9.4-viya-azuredeploy.json": [
 628,
 1447,
-1452,
```
[the remark this answers is not in the document]
Comment (GabinL21) [reply its/ruling/src/integrationTest/resources/expected/azureresourcemanager/azureresourcemanager-S6388.json:9]: It is still raising, just not on the same "level." The 1452 was highlighting the `managedDisk` directly because it was only rising when `managedDisk` existed. 

The issues are now raised on the whole disk (`osDisk`here), just because there are multiple ways of fixing it and we're not sure that `managedDisk`exist when we're raising.

At some point, I tried to have some more logic. Instead of highlighting the whole disk, highlight only the `managed disk` when it already exists. But there are so many edge cases and possibilities that the logic would be way harder to implement.

Same explanation for other similar issues.

File: iac-extensions/arm/src/test/java/org/sonar/iac/arm/tree/impl/bicep/ObjectExpressionImplTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (GabinL21) [reply iac-extensions/arm/src/test/java/org/sonar/iac/arm/tree/impl/bicep/ObjectExpressionImplTest.java]: Ohh ok I didn't see it like that, nice catch!

**Review by jonas-wielage-sonarsource — APPROVED** [review]
LGTM!
